templates/rust/pkgs/default.nix

```diff
@@ -1,6 +1,22 @@
-{rustPlatform}:
+{
+  rustPlatform,
+  # you can add imports here
+}:
 rustPlatform.buildRustPackage rec {
   name = "my-package";
   src = ../.;
-  cargoLock.lockFile = ../Cargo.lock;
+  cargoLock = {
+    lockFile = ../Cargo.lock;
+    # The hash of each dependency that uses a git source must be specified
+    # The hash can be found by setting it to lib.fakeSha256 as shown below and running flox build.
+    # The build will fail but output the expected sha, which can then be added here.
```

Comment:
As @tomberek mentioned the less weird way of getting the hash should be

```
nix store prefetch-file "<url>"
```
However the url part is not necessarily known for cargo dependencies 🤔

Reply:
Will the url be apparent for any dependency a user needs to set the output hash? I feel like we should just recommend the one simplest way

Reply:
no with these generated ones its usually quite difficult to anticipate the url

Reply:
Then I think I would only document the fakeSha256 approach, since this is targeted at a first time user. cc @tomberek

Reply:
Yes, ultimately thats a nix ux issue we won't solve now

Reply:
Tbh, I think it's not the solution we would need in the end, since I definitely do not want to maintain these hashes here manually (those are actually pinned within Cargo.lock). I see that there is an issue (Matthew found that one here: https://github.com/NixOS/nixpkgs/issues/183344) but the provided solution is a workaround for that one (at least from my "simple" user perspective). I therefore suggest to put the actual issue link to the comments as well and somehow record a future todo to revisit / remove the workaround as soon as the nixpkgs issue is solved.

Reply:
👍 tracking the issue of managing the hashes manually

Reply:
Unfortunately https://github.com/NixOS/nixpkgs/issues/183344 seems to be a different issue,but one we have to be aware of regardless.

Reply:
@elbart 
As a bit of a background

Cargo locks git and registry  dependencies differently.

In case of registry provided dependencies (your standard crates.io for example) the lock looks like this:  

```toml
[[package]]
name = "anyhow"
version = "1.0.43"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "28ae2b3dec75a406790005a200b1bd89785afc02517a00ca99ecfe093ee9e6cf"
dependencies = [
 "backtrace",
]
```

Cargo adds a `checksum` to verify the integrity of the downloaded package, this checksum is reused by nix as the hash for this package, hence no need to specify it again.

On the other hand git dependencies are locked like this: 

```toml
[[package]]
name = "libc"
version = "0.2.126"
source = "git+https://github.com/rust-lang/libc?rev=f143f2b9ace1c9b99dd5c104470e94f5990ed7d2#f143f2b9ace1c9b99dd5c104470e94f5990ed7d2"
```

There is no `checksum` provided here as cargo considers the git rev to provide enough integrity guarantees.
Nix does not though. Broadly speaking, any resource fetched by nix needs a checksum.
Nix can't derive it from the lock file so you need to specify it separately.

//

As a side note, (you don't need to look into these yourself these projects might be overwhelming without the nix background)

Alternatively to this pure nix approach, there are external tools that separately, convert your Cargo.toml/Cargo.lock into a nix variant. Since these tools are not part of the nix build they can download and checksum the git dependencies as well as e.g. improve build caching.